api/git/azure.go

```diff
@@ -294,3 +396,189 @@ func getVersionType(name string) string {
 	}
 	return "commit"
 }
+
+func (a *azureClient) listRefs(ctx context.Context, opt option) ([]string, error) {
+	config, err := parseUrl(opt.repositoryUrl)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to parse url")
+	}
+
+	listRefsUrl, err := a.buildRefsUrl(config)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to build list refs url")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", listRefsUrl, nil)
+	if opt.username != "" || opt.password != "" {
+		req.SetBasicAuth(opt.username, opt.password)
+	} else if config.username != "" || config.password != "" {
+		req.SetBasicAuth(config.username, config.password)
+	}
+
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to create a new HTTP request")
+	}
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to make an HTTP request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusNotFound {
+			return nil, ErrIncorrectRepositoryURL
+		} else if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusNonAuthoritativeInfo {
+			return nil, ErrAuthenticationFailure
+		}
+		return nil, fmt.Errorf("failed to list refs url with a status \"%v\"", resp.Status)
+	}
+
+	var refs struct {
+		Value []azureRef
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&refs); err != nil {
+		return nil, errors.Wrap(err, "could not parse Azure refs response")
+	}
+
+	var ret []string
+	for _, value := range refs.Value {
+		if value.Name == "HEAD" {
+			continue
+		}
+		ret = append(ret, value.Name)
+	}
+
+	if a.cacheEnabled {
+		a.repoRefCache.Add(opt.repositoryUrl, ret)
+	}
+	return ret, nil
+}
+
+func (a *azureClient) listFiles(ctx context.Context, opt option) ([]string, error) {
+	var (
+		allPaths    []string
+		filteredRet []string
+		refs        []string
+		err         error
+		success     bool
+		ok          bool
+		cache       interface{}
+	)
+
+	repoKey := generateCacheKey(opt.repositoryUrl, opt.referenceName)
+	if a.repoFileCache != nil {
+		cache, ok = a.repoFileCache.Get(repoKey)
+		if ok {
+			treeCache, success := cache.([]string)
+			if success {
+				for _, path := range treeCache {
+					if matchExtensions(path, opt.extensions) {
+						filteredRet = append(filteredRet, path)
+					}
+				}
+				return filteredRet, nil
+			}
+		}
+	}
+
+	// Check if the reference exists
+	if a.repoRefCache != nil {
+		cache, ok = a.repoRefCache.Get(opt.repositoryUrl)
+		if ok {
+			refs, success = cache.([]string)
+		}
+	}
+	if !ok || !success {
+		opt := option{
+			repositoryUrl: opt.repositoryUrl,
+			username:      opt.username,
+			password:      opt.password,
+		}
+		refs, err = a.listRefs(ctx, opt)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	matchedRef := false
+	for _, ref := range refs {
+		if ref == opt.referenceName {
+			matchedRef = true
+			break
+		}
+	}
+	if !matchedRef {
+		return nil, ErrRefNotFound
+	}
```

Comment:
Why do we even need this? Couldn't we just use `opt.referenceName` and skip all this logic?

Reply:
We can directly use `opt.referenceName`, but it will only show the result until we call the azure listTree(https://docs.microsoft.com/en-us/rest/api/azure/devops/git/trees/get?view=azure-devops-rest-6.0&tabs=HTTP) API at the end. However, in order to using listTree API, we have to know the `{sha1}` value as shown in the API doc which is the `objectID`. This is also partial answer to one of the below comments. As you knew, we need to call `getRootItem` to retrieve the `objectID`. 
So for this snippet of code, we can benefit from circumventing another two API calls if the input reference name doesn't even exist. Also, we can confidently give a clearly error message, like `ErrRefNotFound`.